#!usr/bin/python

'''
Converts the fq file generated by Samtools' consensus calling function into a fasta file
so that we can use it for motif searching
Note this script only keeps the X, 2L, 2R, 3L, and 3R sequences, and discards the rest
'''

import sys

input_file = sys.argv[1]
input_handle = open(input_file, 'r')

output_file = sys.argv[2]
output_handle = open(output_file, 'w')

chrom_names = ['@X', '@2L', '@2R', '@3L', '@3R']

cur_action = "na"
for cur_line in input_handle:
    #Check if we need to update what we're doing
    if cur_line.strip() in chrom_names:
        cur_action = "reading_seq" #Start reading a sequence when we encounter one of the desired chromosomes
        cur_line = cur_line.replace('@', '>') #Replace the @ character with a > character to comply with fasta format
    elif cur_line.strip() == "+":
        cur_action = "reading_qual" #Stop reading a sequence when we encounter a + line [for strand information; in other words, we are done with the last sequence]
    if cur_action == "reading_seq": #If we're reading sequence, then print it to the output file
        output_handle.write(cur_line)

input_handle.close()
output_handle.close()